### PR TITLE
Toponaming/Part: make element transform, _transform and gtransform

### PR DESCRIPTION
**Pull request by bgbsww**

### CODE REVIEW

File: src/Mod/Part/App/TopoShape.h [hunk 1]
```diff
@@ -1313,6 +1361,115 @@ class PartExport TopoShape: public Data::ComplexGeoData
...
+     * @param checkScale: whether to check if the transformation matrix
+     *                    contains scaling factor.
+     * @param copy: whether to perform deep copy of the shape. If noCopy, and
+     *              checkScale, then the shape will be copied if there
+     *              is scaling.
+     *
+     * @return Return a new shape with transformation. The shape itself is not
+     *         modified
+     */
+    TopoShape makeElementTransform(const Base::Matrix4D& mat,
```
Comment (chennes) [inline src/Mod/Part/App/TopoShape.h:1430]: Can this one be `const`?

File: src/Mod/Part/App/TopoShapeExpansion.cpp [hunk 2]
```diff
@@ -2450,6 +2450,107 @@ TopoShape& TopoShape::makeElementOrderedWires(const std::vector<TopoShape>& shap
     return makeElementCompound(wires, nullptr, SingleShapeCompoundCreationPolicy::returnShape);
 }
 
+bool TopoShape::_makeElementTransform(const TopoShape& shape,
+                                      const Base::Matrix4D& rclTrf,
```
Comment (chennes) [inline src/Mod/Part/App/TopoShapeExpansion.cpp:2454]: What is "rclTrf" short for? And it doesn't seem to match the .h file, though the linter is not complaining.

File: src/Mod/Part/App/TopoShape.h (hunk 1 above)
Comment (bgbsww) [reply src/Mod/Part/App/TopoShape.h:1430]: You've lost me here.  The return type?  Making it const is different than all of the other corresponding makeElementX methods, and results in Clang-Tidy: Return type 'const Part::TopoShape' is 'const'-qualified at the top level, which may reduce code readability without improving const correctness.

File: src/Mod/Part/App/TopoShapeExpansion.cpp (hunk 2 above)
Comment (bgbsww) [reply src/Mod/Part/App/TopoShapeExpansion.cpp:2454]: red colored liquid Transformation?  Yeah, that's weird.

File: src/Mod/Part/App/TopoShapeExpansion.cpp [hunk 3]
```diff
@@ -2450,6 +2450,107 @@ TopoShape& TopoShape::makeElementOrderedWires(const std::vector<TopoShape>& shap
...
+
+TopoShape& TopoShape::makeElementTransform(const TopoShape& shape,
+                                           const gp_Trsf& trsf,
+                                           const char* op,
+                                           Copy copy)
+{
+    if (copy == Copy::noCopy) {
+        // OCCT checks the ScaleFactor against gp::Resolution() which is DBL_MIN!!!
+        copy = trsf.ScaleFactor() * trsf.HVectorialPart().Determinant() < 0.
+            || Abs(Abs(trsf.ScaleFactor()) - 1) > Precision::Confusion() ? Copy::copy : Copy::noCopy;
```
[the remark this answers is not in the document]
Comment (bgbsww) [reply src/Mod/Part/App/TopoShapeExpansion.cpp:2478]: Agreed.

**Pull request merged**